### DISCUSSION TITLE
Handling files system in MvvmDialogs

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#14055] [General]
I'm looking into fixing a few things in [HanumanInstitute.MvvmDialogs](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs)

One detail is that file picker InitialFile and InitialFolder were not yet implemented. InitialFile is simple; but InitialFolder? ...

The problem is that I must pass Avalonia type-specific `IStorageFolder`, whereas my code has a generic copy called `IDialogStorageFolder`... maybe I could add another class converting one interface to the other, but that would add an extra layer of complexity when it's already quite complex.

Before implementing a solution, I'd like to have some feedback from experts.

My problem is that most of the MvvmDialogs code has to be platform-agnostic, so I ended up copy/pasting Avalonia's file system classes.

I have these classes so far

[IPathInfoFactory](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs/PathInfo/PathInfoFactory.cs) gives file and folder information, equivalent of Bcl classes in Avalonia. Would be only for desktop, haven't actually used it for a real mobile app yet.

[IDialogStorageFile](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs/FileSystem/DialogStorageFile.cs) and [IDialogStorageFolder](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs/FileSystem/DialogStorageFolder.cs), which are the copies of `IStorageFile` and `IStorageFolder`

One issue I see is that `IDialogStorageFile` must be created using the info from `IPathInfoFactory`; but the later is for desktop only.

The other issue is that I can't convert those classes back to what Avalonia needs.

And finally; the file system is really getting confusing, and I haven't even accounted or tested on mobile devices nor Web browser. Someone is asking what about tracking upload progress in a web browser?

Since this is a critical library for all Avalonia users that is growing in popularity, I'm throwing this out before implementing a solution.

Feedback welcomed!

One solution I implemented elsewhere is like adding an extension method `AsAvalonia` that converts generic classes to platform-specific classes transparently; but that would require creating complex private conversion layers and doesn't address all issues listed above.

**Comment by thevortexcloud** [comment]
Honestly I would either not bother doing this and leave it up to each platform/user to do.  This feels somewhat out of scope for a dialogue framework. But if I had to do it I would make a bunch of platform specific extensions. EG have an Avalonia specific package that implements some interface defined in your shared code/makes some extension methods and then tell your users to use that package. I think that should simplify your code significantly, allow for working with the existing platform interfaces without conversion, and also not require you to reinvent the wheel in a way that every platform can use. This also means your core library can still be platform agnostic and you can also add useful helper methods for other platforms. This is also the approach projects like ReactiveUI uses.

**Reply by mysteryx93** [reply]
The issue is that `OpenFileDialog` is a generic method that is platform-agnostic; thus I can't alter the return type based on what UI you're using.

**Reply by thevortexcloud** [reply]
To be clear, I have not read your code (or used your library). I am just thinking of ways this could be done that does not require you to write a lot of abstractions by delegating the implementation, declaration and use of the file dialogue stuff to the platform specific packages. Where exactly is this method declared? From looking at your code I can't see it. The closest I can see is `ShowOpenFileDialogAsync`.

On a side note, if you have a reference to the actual control (which I think you do?) you should be able to get to the storage provider via the top level https://docs.avaloniaui.net/docs/concepts/toplevel which is a static method.

**Reply by mysteryx93** [reply]
`IStorageProvider` is what's used in the back-end. The goal of MVVM is that the ViewModel doesn't need to rely on UI-specific code (as much as possible).

One option could be to change the `ShowOpenFileDialogAsync` signature to be platform-specific both in its options and its return type. Not sure I want to go that road; and it would bring severe breaking changes. Also, if it's platform-specific types, then you might as well just use `IStorageProvider` directly!

`ShowOpenFileDialogAsync` is defined as extension methods in the shared library; then implemented in `Avalonia.DialogFactory`.

I'm looking at how `IDialogStorageFile` is initiated:
- `PathInfoFactory` and `DialogStorageFile` for desktop
- `BookmarkFileSystem` for mobile
- `AvaloniaDialogStorageFile` to bridge the Avalonia type into the generic type.

It's looking fine except that those usages are all spread out. If I have a `AvaloniaDialogStorageFolder` as start location, I can get the `IStorageFolder` that it wraps, but if it's taken in other ways, then I would need another similar wrapper on top of it...

I also hadn't tested the Browser version and that one requires some work.

In Android and Browser, how does StartLocation typically gets set when looking for a file? How to get the `IStorageFolder` reference usually?

Some things I'm seeing to change.

- In `MvvmDialogs/FileSystem`, the interfaces are generic, but the `DialogStorageFile` implementation is desktop-only, and goes with desktop-only `PathInfoFactory` that is elsewhere. Those 2 classes need to be together and better marked as Desktop-only.
- In `Avalonia/AvaloniaDialogStorageFile`, the file was wrongly named `DialogStorageFile`, renamed it and moved it into `FileSystem` folder. Adding a 2nd wrapper like that should be fine as long as it's clean.

**Comment by mysteryx93** [comment]
Oh geeze. I'm implementing some changes, but running into a stumbling block.

`IStorageItem` is marked as `[NotClientImplementable]`! So I can't create a wrapper around it.

Any way to ignore and bypass that flag?

The only place where that wrapper seems necessary so far is to set `SuggestedStartLocation`.

**Reply by timunie** [reply]
I don't see the reason for you to implement that interface so far. Afaik StorageProvider can create the item from a path or Uri so you could pass that as the start location

**Reply by timunie** [reply]
Maybe also check what I do in Avalonia.Samples to get access to the TopLevel. Esp. the DialogManager may be what you are looking for.

**Reply by mysteryx93** [reply]
If I use this method to convert back to the Avalonia type, will this work on all platforms including iOS and Browser?
`Task<IStorageFolder?> StorageProvider.TryGetFolderFromPathAsync(Uri folderPath);`

**Reply by mysteryx93** [reply]
Would this work reliably?

    internal static async Task<IStorageFolder?> ConvertAsync(this IDialogStorageFolder item)
    {
        if (item is AvaloniaDialogStorageFolder av)
        {
            return av.Source;
        }
        var storage = GetStorageProvider();
        if (storage != null)
        {
            return await storage.TryGetFolderFromPathAsync(item.Path);
        }
        return null;
    }
    
    private static IStorageProvider? GetStorageProvider()
    {
        var result = Application.Current?.ApplicationLifetime switch
        {
            IClassicDesktopStyleApplicationLifetime d => d.MainWindow,
            ISingleViewApplicationLifetime s => s.MainView as TopLevel,
            _ => null
        };
        return result?.StorageProvider;
    }

**Reply by mysteryx93** [reply]
Another question; `IStorageFolder.CreateFileAsync`, what happens if the file already exists? An exception or it overwrites it?

**Comment by mysteryx93** [comment]
Alright I've implemented the changes
https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/releases/tag/v2.1
